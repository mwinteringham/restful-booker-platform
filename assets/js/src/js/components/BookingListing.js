import React, { useEffect, useState } from 'react';
import DatePicker from 'react-datepicker';
import moment from 'moment/moment';
import { API } from '../libs/Api.js';

import "react-datepicker/dist/react-datepicker.css";

const BookingListing = ({booking, getBookings, roomPrice}) => {

    const [allowEdit, toggleEdit] = useState(false);
    const [editBooking, setEditBooking] = useState({});

    useEffect(() => {
        setEditBooking(booking);
    }, [])

    const doDelete = () => {
        API.deleteBooking(editBooking.bookingid, getBookings)
    }

    const doEdit = () => {
        API.updateBooking(editBooking, toggleEdit, getBookings)
    }

    const handleDateChange = (date, target) => {
        editBooking.bookingdates[target] = moment(date.toUTCString()).format("YYYY-MM-DD");;

        setEditBooking(prevState => ({
            ...prevState
        }));
    }

    const updateState = (event) => {
        editBooking[event.name] = event.value;

        setEditBooking(prevState => ({
            ...prevState
        }));
    }

    let calculatedPrice;
    if(roomPrice && booking.bookingdates){
        calculatedPrice = roomPrice * Math.round(Math.abs((new Date(booking.bookingdates.checkin).getTime() - new Date(booking.bookingdates.checkout).getTime())/(24*60*60*1000)));
    } else {
        calculatedPrice = 0;
    }
    
    let bookingView = null;

    if(allowEdit){
        bookingView = <div className="row">
                    <div className="col-sm-2"><input type="text" className="form-control" name="firstname" defaultValue={booking.firstname} onChange={e => updateState(e.target)} /></div>
                    <div className="col-sm-2"><input type="text" className="form-control" name="lastname" defaultValue={booking.lastname} onChange={e => updateState(e.target)} /></div>
                    <div className="col-sm-1"><p>{roomPrice}</p></div>
                    <div className="col-sm-2">
                        <select className="form-control" defaultValue={booking.depositpaid} name="depositpaid" onChange={e => updateState(e.target)}>
                            <option value="false">false</option>
                            <option value="true">true</option>
                        </select>
                    </div>
                    <div className="col-sm-2"><DatePicker className="form-control" selected={moment(booking.bookingdates.checkin).utc(true).toDate()} onChange={date => handleDateChange(date, 'checkin')} dateFormat="yyyy-MM-dd" /></div>
                    <div className="col-sm-2"><DatePicker className="form-control" selected={moment(booking.bookingdates.checkout).utc(true).toDate()} onChange={date => handleDateChange(date, 'checkout')} dateFormat="yyyy-MM-dd" /></div>
                    <div className="col-sm-1">
                        <span className="fa fa-check confirmBookingEdit" onClick={doEdit} style={{paddingRight : 10 + "px"}}></span>
                        <span className="fa fa-remove exitBookingEdit" onClick={() => {toggleEdit(false)}}></span>
                    </div>
                </div>
    } else {
        bookingView = <div className="row">
                    <div className="col-sm-2"><p>{booking.firstname}</p></div>
                    <div className="col-sm-2"><p>{booking.lastname}</p></div>
                    <div className="col-sm-1"><p>{calculatedPrice}</p></div>
                    <div className="col-sm-2"><p>{String(booking.depositpaid)}</p></div>
                    <div className="col-sm-2"><p>{booking.bookingdates.checkin.split('T')[0]}</p></div>
                    <div className="col-sm-2"><p>{booking.bookingdates.checkout.split('T')[0]}</p></div>
                    <div className="col-sm-1">
                        <span className="fa fa-pencil bookingEdit" onClick={() => {toggleEdit(true)}} style={{paddingRight: 10 + "px"}}></span>
                        <span className="fa fa-trash bookingDelete" onClick={doDelete}></span>
                    </div>
                    </div>
    }

    return(
        <div className={"detail booking-" + booking.roomid}>
            {bookingView}
        </div>
    )

}

export default BookingListing;
