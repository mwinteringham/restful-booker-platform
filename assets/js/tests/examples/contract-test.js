import React from 'react';
import {
    render,
    fireEvent,
    cleanup
  } from '@testing-library/react'
import BookingListing from '../../src/js/components/BookingListing.js';
import contract from '../../../../booking/src/test/resources/contract.json';

// We first declare our test by using Jests test function and provide it
// with a name and then an anonymous function which will run our test
test('Rooms list component', () => {

    // First we create our component that we want to check. By using
    // react testing libraries' render function and passing it the 
    // React component we want to create, RTL will create a headless
    // DOM and render the React component inside. 
    //
    // You'll notice we are passing parameters to the component namely 
    // the contract we imported at the top of test. If the contract 
    // has been changed to make a different check pass we expect the 
    // check to fail and show what's changed.
    //
    // Also we pass isAuthenticated to put it into a mode where it 
    // renders as if we were logged in.
    const booking = render(
        <BookingListing booking={contract} isAuthenticated={true}/>
    );

    // We then compare what the component has done to a previously saved
    // state that lives in the __snapshots__ folder. If anything has changed
    // it will raise an alert.
    expect(booking).toMatchSnapshot();
});