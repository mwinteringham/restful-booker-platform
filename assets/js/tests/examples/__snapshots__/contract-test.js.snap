// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rooms list component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="detail booking-1"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-2"
          >
            <p>
              James
            </p>
          </div>
          <div
            class="col-sm-2"
          >
            <p>
              Dean
            </p>
          </div>
          <div
            class="col-sm-1"
          >
            <p>
              0
            </p>
          </div>
          <div
            class="col-sm-2"
          >
            <p>
              true
            </p>
          </div>
          <div
            class="col-sm-2"
          >
            <p>
              2022-02-01
            </p>
          </div>
          <div
            class="col-sm-2"
          >
            <p>
              2022-02-05
            </p>
          </div>
          <div
            class="col-sm-1"
          >
            <span
              class="fa fa-pencil bookingEdit"
              style="padding-right: 10px;"
            />
            <span
              class="fa fa-trash bookingDelete"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="detail booking-1"
    >
      <div
        class="row"
      >
        <div
          class="col-sm-2"
        >
          <p>
            James
          </p>
        </div>
        <div
          class="col-sm-2"
        >
          <p>
            Dean
          </p>
        </div>
        <div
          class="col-sm-1"
        >
          <p>
            0
          </p>
        </div>
        <div
          class="col-sm-2"
        >
          <p>
            true
          </p>
        </div>
        <div
          class="col-sm-2"
        >
          <p>
            2022-02-01
          </p>
        </div>
        <div
          class="col-sm-2"
        >
          <p>
            2022-02-05
          </p>
        </div>
        <div
          class="col-sm-1"
        >
          <span
            class="fa fa-pencil bookingEdit"
            style="padding-right: 10px;"
          />
          <span
            class="fa fa-trash bookingDelete"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
